File: src/sim/SpacePartitioningOptions.py
##This object is used to pass value to global variable intervalsize used in src/sim/space_partitioning.cc. 


from m5.SimObject import SimObject 
from m5.params import *

class SpacePartitioningOptions(SimObject):
    type = 'SpacePartitioningOptions'
    cxx_header = 'sim/space_partitioning_options.hh'
    cxx_class="gem5::SpacePartitioningOptions"
    algo = Param.Int('0', 'algo to use in space partitioning');
    num_cores = Param.Int('8', "Number of cores being recorded");
    num_llcs = Param.Int('2', "Number of LLCs to record");
    interval_size = Param.UInt64('5000000', "Space partitioinig interval");
    l2_assoc = Param.Int('16', "LLC associativity");
    dss_size = Param.Int('64', "DSS size of UMON");
    stack_size = Param.Int('16', "Max number of addresses stored for a set in DSS");
    learning_rate = Param.Float('0.8', "Learning rate");
    discount = Param.Float('0.1',"discount rate in Q learning");
    private_ways = Param.Int('12', "Number of ways that are private to cores, rest will be shared always");
    q_file_path = Param.String("", 'The file path for the q_table to be loaded and saved to');
    l2_size = Param.UInt64('4194304', '1 LLC slice size');
    max_classes = Param.Int('16', 'Maximum number of classes in CPA');
    icov_thresh = Param.Float('0.2', 'interval coeff of variation thesh, used for phase change detection');

File: src/sim/space_partitioning_options.cc
#include "sim/space_partitioning_options.hh"

#include<iostream>

#include "base/trace.hh"
#include "base/statistics.hh"

namespace gem5
{

SpacePartitioningOptions::SpacePartitioningOptions(const SpacePartitioningOptionsParams &p) : 
	SimObject(p),
	algo(p.algo),
	interval_size(p.interval_size),
	num_cores(p.num_cores),
    num_llcs(p.num_llcs),
	l2_assoc(p.l2_assoc),
	dss_size(p.dss_size),
	stack_size(p.stack_size),
	lr(p.learning_rate),
	discnt(p.discount),
	prvt_ways(p.private_ways),
	q_file_path(p.q_file_path),
	l2_size(p.l2_size),
	max_classes(p.max_classes),
	icov_thresh(p.icov_thresh)
{
	SpacePartitioning::setSpacePOptions((int)algo, (uint64_t)interval_size,
	(int)num_cores, (int)num_llcs, (int)l2_assoc, (int) dss_size, (int) stack_size,
	(double)lr, (double)discnt, (int)prvt_ways, q_file_path,
	(uint64_t)l2_size, (int)max_classes, (double)icov_thresh );
}
}//namespace gem5

